src/ray/gcs/gcs_server/gcs_task_manager.h

```diff
@@ -181,36 +181,46 @@ class GcsTaskManager : public rpc::TaskInfoHandler {
     std::vector<rpc::TaskEvents> GetTaskEvents(
         const absl::flat_hash_set<TaskAttempt> &task_attempts) const;
 
-    ///  Mark tasks from a job as failed.
+    ///  Mark tasks from a job as failed as job ends.
     ///
     /// \param job_id Job ID
     /// \param job_finish_time_ns job finished time in nanoseconds, which will be the task
     /// failed time.
-    void MarkTasksFailed(const JobID &job_id, int64_t job_finish_time_ns);
+    void MarkTasksFailedOnJobEnds(const JobID &job_id, int64_t job_finish_time_ns);
 
    private:
     /// Mark the task tree containing this task attempt as failure if necessary.
     ///
-    /// A task tree is represented by the `parent_task_id` field for each task.
-    ///
-    /// All non-terminated (not finished nor failed) tasks in the task tree with the task
-    /// event as the root should be marked as failed with the root's failed timestamp if
-    /// any of:
-    ///     1. The task itself fails.
-    ///     2. Its parent failed, and neither itself nor its parent is detached actor
-    ///     task.
-    ///     3. Its parent failed, and both itself and its parent are detached actor task.
-    /// TODO(rickyx): Detached actors tasks are now currently marked as failed if parent
-    /// dies.
+    /// See `ShouldMarkTaskTreeFailed` for details on when a task tree should be marked as
+    /// failed.
     ///
     /// NOTE: Since there is delay in task events reporting, a task's state might not be
     /// reported before a worker is killed, therefore, some tasks might have actually
     /// run/been submitted but it's states not being reported to GCS. This is hard to
     /// resolve unless we add synchronized reporting before submitting a task.
     ///
     /// \param task_id ID of the task event that's the root of the task tree.
-    /// \param parent_task_id ID of the task's parent.
-    void MarkTaskTreeFailedIfNeeded(const TaskID &task_id, const TaskID &parent_task_id);
+    void MarkTaskTreeFailedIfNeeded(const TaskID &task_id);
+
+    /// Check if the child task should be marked as failed.
+    ///
+    /// A child task should be marked as failed iff:
+    ///   0. The child task is still running AND
+    ///   1. The parent task failed with errors where worker is killed
+    ///     (e.g. ACTOR_DIED, WORKER_DIED) AND
+    ///   2. The child task is NOT from a detached actor.
+    ///
+    /// For other cases, such as a parent task failed with application exceptions,
+    /// Ray currently doesn't kill the child tasks. Therefore, we don't mark the child as
+    /// failed.
+    ///
+    /// \param child_task The child task.
+    /// \param parent_task The parent task.
+    /// \param failed_ts The failure timestamp.
```

Comment:
Can you just take out failed_ts from this method and get it outside this method? Personally don't like this sort of funcs that modify the values inside... 

If not possible, we should document this variable better; It is the [in] param and set when the child should be marked as failed

Reply:
sure thing.